plugin_ide.ui/src/com/github/rustdt/ide/ui/editor/RustEditorActionContributor.java

```diff
@@ -28,6 +51,99 @@ protected RustOpenDefinitionOperation createOpenDefinitionOperation(ITextEditor
 	
 	@Override
 	protected void registerOtherEditorHandlers() {
+		setupOpenType();
+	}
+	
+	private void setupOpenType() {
+		evaluateGlobalSourceStructure();
+		listenToUpdatesOfGlobalSourceStructure();
+	}
+	
+	private void evaluateGlobalSourceStructure() {
+		new Job("Searching types") {
+			@Override
+			protected IStatus run(IProgressMonitor monitor) {
+				try {
+					ResourcesPlugin.getWorkspace().getRoot().accept(RustEditorActionContributor.this::addResource);
+				} catch(Exception e) {
+					AbstractLangCore.log().logError("Global source structure could not be determined", e);
+				}
+				return Status.OK_STATUS;
+			}
+		}.schedule();
+	}
+	
+	private boolean addResource(IResource resource) {
+		convertToRustFileLocation(resource).ifPresent(RustEditorActionContributor::fileTouched);
+		return true;
+	}
+	
+	private void listenToUpdatesOfGlobalSourceStructure() {
+		ResourcesPlugin.getWorkspace().addResourceChangeListener(this::processEvent, IResourceChangeEvent.POST_CHANGE);
 	}
 	
+	private void processEvent(IResourceChangeEvent event) {
+		new Job("Updating types") {
+			@Override
+			protected IStatus run(IProgressMonitor monitor) {
+				try {
+					event.getDelta().accept(RustEditorActionContributor.this::applyResourceDelta);
+				} catch(Exception e) {
+					AbstractLangCore.log().logError("Global source structure could not be determined", e);
+				}
+				return Status.OK_STATUS;
+			}
+		}.schedule();
+	}
+	
+	private boolean applyResourceDelta(IResourceDelta delta) {
+		convertToRustFileLocation(delta.getResource()).ifPresent(location -> {
+			switch(delta.getKind()) {
+			case IResourceDelta.ADDED:
+			case IResourceDelta.CHANGED:
+				fileTouched(location);
+				break;
+			case IResourceDelta.REMOVED:
+				fileRemoved(location);
+				break;
+			}

```

Comment:
Yes, you can't analyze the delta object in a separate thread. The delta object has to be analyzed in the listener thread (the thread that triggers the resource listener). But what I was saying is that the listener analysis should consist only of marking locations as having out-of-date structure/index. You don't actually calculate the new structure/index in the listener thread. Rather, when the index is marked as out of date for a location, that triggers a recalculation in a background thread. Again the DocumentReconcileManager would be a good example of a such a design. But it is generally more or less what you said, scheduling one job per file/location. (with a few more minor details)